File: othello/fixtures.py
"""Starting positions for play and tests."""
from __future__ import annotations
from othello.core import Board, BLACK, FULL, square_to_move

def init_early_game() -> Board:
    BLACK_START = square_to_move('D5') | square_to_move('E4')
    WHITE_START = square_to_move('D4') | square_to_move('E5')
    return Board(black=BLACK_START, white=WHITE_START, to_move=BLACK)

def init_near_terminal_game_white_win() -> Board:
    EMPTY_START = (
        square_to_move("A1")
        | square_to_move("E1")
        | square_to_move("A2")
        | square_to_move("B7")
        | square_to_move("A8")
        | square_to_move("B8")
    )
    BLACK_START = (
        square_to_move("F1")
        | square_to_move("E2")
        | square_to_move("D3")
        | square_to_move("F3")
        | square_to_move("C4")
        | square_to_move("F4")
        | square_to_move("B5")
        | square_to_move("F5")
        | square_to_move("C6")
        | square_to_move("C7")
        | square_to_move("D7")
        | square_to_move("E7")
        | square_to_move("C8")
        | square_to_move("D8")
        | square_to_move("E8")
    )
    WHITE_START = FULL ^ BLACK_START ^ EMPTY_START
    return Board(black=BLACK_START, white=WHITE_START, to_move=BLACK)

def init_near_terminal_game_black_win() -> Board:
    BLACK_START = 0x0DEFEF236143030F
    WHITE_START = 0x9010105C9CBCFCE0
    return Board(black=BLACK_START, white=WHITE_START, to_move=BLACK)

def init_near_terminal_game_either_can_win() -> Board:
    BLACK_START = 0x0DEFEF236143032F
    WHITE_START = 0x9010105C9CBCFCC0
    return Board(black=BLACK_START, white=WHITE_START, to_move=BLACK)

# Named starts for the CLI (--start).
STARTS = {
    "early": init_early_game,
    "white-win": init_near_terminal_game_white_win,
    "black-win": init_near_terminal_game_black_win,
    "either": init_near_terminal_game_either_can_win,
}

File: rust/src/fixtures.rs
//! Starting positions for play and tests, port of `fixtures.py`.

use crate::core::{square_to_move, Board, BLACK, FULL};

fn sq(name: &str) -> u64 {
    square_to_move(name).expect("valid fixture square")
}

pub fn init_early_game() -> Board {
    let black = sq("D5") | sq("E4");
    let white = sq("D4") | sq("E5");
    Board::new(black, white, BLACK)
}

pub fn init_near_terminal_game_white_win() -> Board {
    let empty = sq("A1") | sq("E1") | sq("A2") | sq("B7") | sq("A8") | sq("B8");
    let black = sq("F1")
        | sq("E2")
        | sq("D3")
        | sq("F3")
        | sq("C4")
        | sq("F4")
        | sq("B5")
        | sq("F5")
        | sq("C6")
        | sq("C7")
        | sq("D7")
        | sq("E7")
        | sq("C8")
        | sq("D8")
        | sq("E8");
    let white = FULL ^ black ^ empty;
    Board::new(black, white, BLACK)
}

pub fn init_near_terminal_game_black_win() -> Board {
    Board::new(0x0DEF_EF23_6143_030F, 0x9010_105C_9CBC_FCE0, BLACK)
}

pub fn init_near_terminal_game_either_can_win() -> Board {
    Board::new(0x0DEF_EF23_6143_032F, 0x9010_105C_9CBC_FCC0, BLACK)
}

/// Named starts for the CLI (--start), sorted (matches Python's `sorted(STARTS)`).
pub const START_NAMES: [&str; 4] = ["black-win", "either", "early", "white-win"];

pub fn start(name: &str) -> Option<Board> {
    match name {
        "early" => Some(init_early_game()),
        "white-win" => Some(init_near_terminal_game_white_win()),
        "black-win" => Some(init_near_terminal_game_black_win()),
        "either" => Some(init_near_terminal_game_either_can_win()),
        _ => None,
    }
}
